pkg/pgclient/config.go

```diff
@@ -154,32 +158,76 @@ func (cfg *Config) GetConnectionStr() string {
 	return cfg.DbUri
 }
 
-// GetPoolSize returns the max pool size based on the max_connections allowed by database and the defaultFraction.
-// Arg inputPoolSize is the pool size provided in the CLI flag.
-func (cfg *Config) GetPoolSize(poolName string, defaultFraction float64, inputPoolSize int) (int, error) {
-	maxConns, err := cfg.maxConn()
-	if err != nil {
-		return 0, fmt.Errorf("max connections: %w", err)
+// GetPoolSizes returns the pool sizes adjusted according to defaults, read only settings, and max connections available
+func (cfg *Config) GetPoolSizes(readOnly bool, dbMaxConns int) (readerPoolSize, writerPoolSize, maintPoolSize int, err error) {
+	cfgMaxConns := cfg.MaxConnections
+	if cfgMaxConns == defaultMaxConns {
+		cfgMaxConns = int(float64(dbMaxConns) * 0.8)
+	}
+	if cfgMaxConns > dbMaxConns {
+		log.Warn("msg", "configured db.connections-max is greater then postgres max_connections",
+			"db.connections-max", cfgMaxConns,
+			"max_connections", dbMaxConns,
+		)
 	}
 
-	switch {
-	case inputPoolSize == defaultPoolSize:
-		// For the default case, we need to take up defaultFraction of allowed connections.
-		poolSize := float64(maxConns) * defaultFraction
-		if cfg.UsesHA {
-			poolSize /= 2
+	calc := func(inputPoolSize int, fraction float64) int {
+		if inputPoolSize == defaultPoolSize {
+			poolSize := float64(dbMaxConns) * fraction
+			if cfg.UsesHA {
+				poolSize /= 2
+			}
+			return int(poolSize)
 		}
-		return int(poolSize), nil
-	case inputPoolSize < MinPoolSize:
-		return 0, fmt.Errorf("%s pool size canot be less than %d: received %d", poolName, MinPoolSize, inputPoolSize)
-	case inputPoolSize > maxConns:
-		return 0, fmt.Errorf("%s pool size canot be greater than the 'max_connections' allowed by the database", poolName)
-	default:
+		return inputPoolSize
+	}
+
+	// calc maint pool size
+	maintPoolSize = cfg.MaintPoolSize
+	if maintPoolSize == defaultPoolSize {
+		maintPoolSize = defaultMaintPoolSize
+	}
+
+	// calc writer pool size
+	if !readOnly {
+		writerPoolSize = calc(cfg.WriterPoolSize, defaultConnFraction)
+	}
+
+	// calc reader pool size
+	readerFraction := defaultReaderFraction
+	if readOnly {
+		readerFraction = defaultConnFraction
+	}
+	readerPoolSize = calc(cfg.ReaderPoolSize, readerFraction)
+	if cfg.ReaderPoolSize == defaultPoolSize && readerPoolSize+writerPoolSize+maintPoolSize > cfgMaxConns {
+		readerPoolSize = readerPoolSize - maintPoolSize
+	}
```

Comment:
I'd split this off in its own function that returns the pool sizes to contain the logic.

Reply:
i moved basically everything except this into other functions, so I think it's okay to leave it here. okay?